packages/workflow/src/internals.ts

```diff
@@ -534,14 +542,96 @@ export class Activator implements ActivationHandler {
     );
   }
 
+  public doUpdate(activation: coresdk.workflow_activation.IDoUpdate): void {
+    const { id, name, headers, runValidator } = activation;
+    if (!id) {
+      throw new TypeError('Missing activation update id');
+    }
+    if (!name) {
+      throw new TypeError('Missing activation update name');
+    }
+    if (!this.updateHandlers.has(name)) {
+      // TODO (dan): Signal is able to handle this situation more gracefully by
```

Comment:
I don't think the way we handle the signal case would be applicable to updates. Validation has to return immediately, no?

One thing that could make sense however would be that if an update handler doesn't exist yet, we postpone it as much as possible inside the scope of the current activation (arguably, inside the scope of the WFT would be even better, but there's currently no easy way to know if an activation is the last one for this WFT). I thikn that would cover many practical cases.

Still, another problem with how signal get dispatched immediately or buffered: there is no guarantee that signals will get dispatched in their history order. For example, if a workflow receive signal A then B, both before the very first WFT get accepted by a worker, and the workflow does `setHandler(B, ...)` then `setHandler(A, ...)`, then signal B will indeed get dispatched first, which is very surprising for users and almost always wrong. We'd want to fix that at some point, but that's another discussion. The real point here is that I don't think that the way we "gracefully buffer signals" is exactly the thing we'd want to replicate with updates.

Reply:
:+1: We want to reject updates immediately if there is nothing that can handle them. I think it's a given that update is not like signal. TODO can be removed IMO, behavior as written is correct.

Reply:
I'm not saying we don't want to do some "graceful" handling, just stating that the way gracefully handling signals is fundamentally flawed and definitely not something we want to reproduce.

The problem is that handlers can't be registered before the main workflow function starts executing, which means that signals and update sent before the first WFT Started would automatically get discarded (for signals) or refused (for updates). Delaying dispatch of those later in the activation (ie. after the main function is given a chance to execute) resolves the problem, without compromising the fact that the validator must complete in the same WFT.

I think that's the only "buffering"/"delaying" that we really need.

Reply:
:+1: Forgot TS requires manual handler registration. So Go had this problem too: https://github.com/temporalio/sdk-go/blob/0352634b45b1b81293c9bf3cec3bb033b82f841f/internal/internal_update.go#L263-L271. Yeah one tick of the event loop can hopefully do it.

Reply:
@dandavison Search for `workflowFunctionWasCalled` in this file to see how we handle the same thing for queries. I think how we deal with queries in that regard is snaer than what we do for signals.

Reply:
Looks like you figured it out, we need a tick of the event loop before failing an update to allow late registration.

Reply:
Can someone give me more of a hint here? I thought that perhaps what you were saying was that this would work if we scheduled the continuation of `doUpdate` as a promise. However, I've done that in this commit, and added tests for late registration of signal and update handlers. The signal test passes as expected (and fails if I change the code not to buffer signals), but the update test times out (never finds the late-regsitered handler) when the `doUpdate` continuation is scheduled as a promise.

Reply:
[compare/sdk-879-workflow-update...dandavison:temporalio-sdk-typescript:sdk-879-workflow-update-late-registration](https://github.com/dandavison/temporalio-sdk-typescript/compare/sdk-879-workflow-update...dandavison:temporalio-sdk-typescript:sdk-879-workflow-update-late-registration?expand=1)

Reply:
I had to go before I could finish this review but you need to wait for the workflow function to be evaluated before you reject this update.

Reply:
Please address this before you merge including the relevant tests.

Reply:
> I had to go before I could finish this review but you need to wait for the workflow function to be evaluated before you reject this update.

It's like Fermat's Last Theorem :)